Add basic support for access control headers to ActionDispatch::Static

yuki24 opened the pull request:
Now ActionDispatch::Static can accept access control headers such as `Access-Control-Allow-Origin`. These headers will be set as well as `Cache-Control` header when a response is delivered. Access control headers can be configured with `#config` in `config/application.rb` or `config/environments/***.rb`:

``` ruby
config.action_dispatch.access_control_headers = {
  "Access-Control-Allow-Origin"      => 'http://rubyonrails.org',
  "Access-Control-Expose-Headers"    => 'X-Other-Custom-Header',
  "Access-Control-Allow-Headers"     => 'X-Another-Custom-Header',
  "Access-Control-Max-Age"           => '42',
  "Access-Control-Allow-Credentials" => 'true',
  "Access-Control-Allow-Methods"     => 'GET'
}
```

cc/ @schneems

yuki24 commented on actionpack/lib/action_dispatch/middleware/static.rb:
```diff
@@ -13,10 +16,10 @@ module ActionDispatch
   # does not exist a 404 "File not Found" response will be returned.
   class FileHandler
-    def initialize(root, cache_control)
+    def initialize(root, cache_control, access_control_headers)
```
Although this implementation works, I'm not sure about this interface since `Cache-Control` is also part of the HTTP headers and the third argument seems redundant. I guess this interface could be just`default_headers`.

yuki24 commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -17,6 +17,7 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_dispatch.encrypted_cookie_salt = 'encrypted cookie'
     config.action_dispatch.encrypted_signed_cookie_salt = 'signed encrypted cookie'
     config.action_dispatch.perform_deep_munge = true
+    config.action_dispatch.access_control_headers = { }
```
`access_control_headers` will directly be passed into `Rack::File`, so the name doesn't necessarily have to be prefixed by `access_control_`. But if we change this name to just `asset_headers` or `static_file_headers`, then we also should deprecate `static_cache_control` config in favor of it.

schneems commented:
Ultimately I think it looks good. Let's change some naming so that it's a little more generic. Instead of `access_control_headers` let's just use `headers` which is technically what it is (although you can use them as access control headers).

You can reduce the initialization logic to something like this:

```
headers ||= {}
headers['Cache-Control'] = cache_control
```

I would also change `config.action_dispatch.access_control_headers` to `config.action_dispatch.asset_headers` or something more accurate. We can then explicitly add a section in the guides that shows how to set access control headers  http://guides.rubyonrails.org/asset_pipeline.html

I agree we should deprecate some interfaces here and clean some things up. Unfortunately as I understand, the deadline for getting deprecations in to Rails 5 was Rails 4.2.0

yuki24 commented:
Updated the commit as suggested. Thanks @schneems for reviewing!

yuki24 commented:
forgot to add a CHANGELOG entry. I'll add it tonight.

schneems commented on actionpack/lib/action_dispatch/middleware/static.rb:
```diff
@@ -13,11 +16,12 @@ module ActionDispatch
-      @file_server = ::Rack::File.new(@root, headers)
+
+      headers ||= {}
+      @file_server = ::Rack::File.new(@root, headers.reverse_merge('Cache-Control' => cache_control))
```
`merge` and `reverse_merge` are really expensive. Even though this only gets called once on initialization the code in this file is very performance sensitive and I don't want people copying existing style. Can you also factor out the merge?

yuki24 replied:
Thanks @schneems, removed the merge call,

jeremy commented on actionpack/lib/action_dispatch/middleware/static.rb:
```diff
@@ -94,9 +98,9 @@ def gzip_file_path(path)
   # requests will result in a file being returned.
   class Static
-    def initialize(app, path, cache_control=nil)
+    def initialize(app, path, cache_control = nil, headers = {})
```
Use keyword args for newly added arguments: `headers: nil`

yuki24 replied:
:+1: replaced with `headers: {}` kwarg.

jeremy commented on actionpack/lib/action_dispatch/middleware/static.rb:
```diff
@@ -94,9 +98,9 @@ def gzip_file_path(path)
+    def initialize(app, path, cache_control = nil, headers = {})
       @app = app
-      @file_handler = FileHandler.new(path, cache_control)
+      @file_handler = FileHandler.new(path, cache_control, headers)
```
Since we're passing response headers to FileHandler now, we can deprecate the `cache_control` argument and handle it here rather than in `FileHandler#initialize`:
```ruby
# Static
def initialize(app, path, deprecated_cache_control = :not_set, headers: nil)
  if deprecated_cache_control != :not_set
    ActiveSupport::Deprecation.warn "The `cache_control = '…'` argument is deprecated, replaced by `headers: { 'Cache-Control' => '…' }`, and will be removed in Rails 5.1."
    headers = (headers || {}).merge('Cache-Control'.freeze => deprecated_cache_control)
  end

  @app = app
  @file_handler = FileHandler.new(path, headers: headers)
end

# FileHandler
def initialize(root, headers:)
  …
  @file_server = ::Rack::File.new(@root, headers)
end
```

jeremy replied:
(That'd mean deprecating `config.static_cache_control` as well, transitioning to a headers config setting.)

yuki24 replied:
changed it as suggested.

jeremy commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -17,6 +17,7 @@ class Railtie < Rails::Railtie # :nodoc:
     config.action_dispatch.encrypted_cookie_salt = 'encrypted cookie'
     config.action_dispatch.encrypted_signed_cookie_salt = 'signed encrypted cookie'
     config.action_dispatch.perform_deep_munge = true
+    config.action_dispatch.asset_headers = { }
```
This sounds like headers on the asset pipeline rather than headers on static file response.

The other config scattered among `config.static_cache_control` and `config.serve_static_files`. Let's unify these as "public file server" config, reflecting that we're just serving files from the highly visible `public/` dir:
```ruby
config.public_file_server.enabled = true  # replace config.serve_static_files
config.public_file_server.headers = { 'Cache-Control' => '…' }  # replace cache.static_cache_control
```

yuki24 replied:
added `config.public_file_server` namespace, but will send a separate pull request for adding `config.public_file_server.enabled`.

jeremy commented on actionpack/test/dispatch/static_test.rb:
```diff
@@ -143,6 +152,17 @@ def test_serves_gzip_with_propper_content_type_fallback
+    assert_equal 'true', response.headers["Access-Control-Allow-Credentials"]
+    assert_equal 'GET',  response.headers["Access-Control-Allow-Methods"]
+  end
+
```
Calling out Access-Control headers suggests that we built a feature to add these headers in particular. Using arbitrary/meaningless headers would make it clearer that we support any response headers.

yuki24 replied:
added `Cache-Control` and a custom header :+1:

jeremy commented on actionpack/test/dispatch/static_test.rb:
```diff
@@ -199,7 +219,7 @@ class StaticTest < ActiveSupport::TestCase
   def setup
     @root = "#{FIXTURE_LOAD_PATH}/public"
-    @app = ActionDispatch::Static.new(DummyApp, @root, "public, max-age=60")
+    @app = ActionDispatch::Static.new(DummyApp, @root, "public, max-age=60", ACCESS_CONTROL_HEADERS)
```
These headers are not needed for all tests. Would instantiate `@app = …` in the specific test case instead.

yuki24 replied:
done :+1:

jeremy commented on actionpack/test/dispatch/static_test.rb:
```diff
@@ -228,7 +248,7 @@ def test_custom_handler_called_when_file_is_outside_root
   def setup
     @root = "#{FIXTURE_LOAD_PATH}/公共"
-    @app = ActionDispatch::Static.new(DummyApp, @root, "public, max-age=60")
+    @app = ActionDispatch::Static.new(DummyApp, @root, "public, max-age=60", ACCESS_CONTROL_HEADERS)
```
Ditto.

jeremy commented on railties/lib/rails/application/default_middleware_stack.rb:
```diff
@@ -18,7 +18,7 @@ def build_stack
           if config.serve_static_files
-            middleware.use ::ActionDispatch::Static, paths["public"].first, config.static_cache_control
+            middleware.use ::ActionDispatch::Static, paths["public"].first, config.static_cache_control, config.action_dispatch.asset_headers
           end
```
Will need to use `config.public_file_server.*` config here, and introduce deprecations either on `Rails::Application::Configuration#serve_static_files=` and `#static_cache_control=` (they're attr accessors now), or handle deprecations here, where the config is read. Probably better to get a deprecation warning when they're set rather than when they're read.

yuki24 replied:
Added a deprecation in the setter, I'm not sure if it looks good enough, though. I'll send a separate pull request for replacing `config.serve_static_files` with `config.public_file_server.enabled`.

yuki24 commented on actionpack/lib/action_dispatch/railtie.rb:
```diff
@@ -24,6 +24,9 @@ class Railtie < Rails::Railtie # :nodoc:
     }
 
+    config.public_file_server = ActiveSupport::OrderedOptions.new
+    config.public_file_server.headers = { 'Cache-Control' => 'public, max-age=3600' }
```
I'm not sure if this is the right place to define `public_file_server`. I'm open to suggestions.

yuki24 commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -51,6 +51,14 @@ def initialize(*)
+                                        "Please use `config.public_file_server.headers['Cache-Control'] = #{value} instead.")
+
+        @static_cache_control = value
+        public_file_server.headers['Cache-Control'.freeze] = value
```
This is a bit smelly, but somewhere we have to this to avoid duplicate warnings from `ActionDispatch::Static`.

yuki24 committed:
Add the ability of returning arbitrary headers to ActionDispatch::Static

Now ActionDispatch::Static can accept HTTP headers so that developers
will have control of returning arbitrary headers like
'Access-Control-Allow-Origin' when a response is delivered. They can
be configured through `#config.public_file_server.headers`:

  config.public_file_server.headers = {
    "Cache-Control"               => "public, max-age=60",
    "Access-Control-Allow-Origin" => "http://rubyonrails.org"
  }

Also deprecate `config.static_cache_control` in favor of
`config.public_file_server.headers`.

schneems commented:
What are we waiting on for this to go in? Anything a blocker here?

yuki24 commented on railties/lib/rails/application/default_middleware_stack.rb:
```diff
@@ -18,7 +18,10 @@ def build_stack
+            headers = config.public_file_server.headers
+            headers['Cache-Control'] = config.static_cache_control if config.static_cache_control.present?
+
+            middleware.use ::ActionDispatch::Static, paths["public"].first, :not_set, config.static_index, headers: headers
```
I needed to use `:not_set` explicitly due to this change: https://github.com/rails/rails/pull/20017. I think we should make the `index` arg a keyword argument so we don't have to care about the order of the args here. Alternatively, we can just merge the `static_index` config into the `public_file_server` namespace, which I actually prefer, but I'm open to suggestions.

cc/ @eliotsykes @jonatack

vipulnsward commented on actionpack/lib/action_dispatch/middleware/static.rb:
```diff
@@ -13,15 +13,13 @@ module ActionDispatch
   # does not exist, a 404 "File not Found" response will be returned.
   class FileHandler
-    def initialize(root, cache_control, index = 'index')
+    def initialize(root, index = 'index', headers: )
```
`index` can be kwarg too.

yuki24 replied:
Absolutely. See my comment above about it: https://github.com/rails/rails/pull/19135#discussion_r32110510

yuki24 replied:
sent a separate pull request for this: https://github.com/rails/rails/pull/20520.

jonatack replied:
The comment states `optional HTTP headers` but the headers keyword is required.

`headers:` are optional or required?

jonatack replied:
Perhaps `headers:` needs a default value.

yuki24 replied:
Headers are totally optional. I changed it to use a `{}` by default.

jonatack replied:
Great.

This PR breaks the interface for other services/gems that hook into it, for instance the Heroku-deflator gem [here](https://github.com/romanbsd/heroku-deflater/blob/master/lib/heroku-deflater/serve_zipped_assets.rb#L15) but seems unavoidable for adding this feature.

yuki24 replied:
oh, I didn't know that `ActionDispatch::FileHandler` is a public class. I don't think it's ok to break it.

yuki24 replied:
I did a little investigation but I'm not sure if we should treat `ActionDispatch::FileHandler` as a public class. It's basically an inner class. Does it still have to be public?

yuki24 commented:
@schneems updated the commit. I'm a bit concerned about the arguments of the `ActionDispatch::Static` initializer, but everything else should be good now.

vipulnsward commented on railties/lib/rails/application/configuration.rb:
```diff
@@ -52,6 +52,13 @@ def initialize(*)
 
+      def static_cache_control=(value)
+        ActiveSupport::Deprecation.warn("static_cache_control is deprecated and will be removed in Rails 5.1. " \
+                                        "Please use `config.public_file_server.headers['Cache-Control'] = #{value} instead.")
```
Would be good to have this as-
```ruby
config.public_file_server.headers= {'Cache-Control' => #{value}}
```
This would suggest the options being passed to headers have been expanded.

yuki24 replied:
updated.

vipulnsward commented on railties/lib/rails/application/default_middleware_stack.rb:
```diff
@@ -18,7 +18,10 @@ def build_stack
           if config.serve_static_files
-            middleware.use ::ActionDispatch::Static, paths["public"].first, config.static_cache_control, config.static_index
+            headers = config.public_file_server.headers
+            headers['Cache-Control'] = config.static_cache_control if config.static_cache_control.present?
```
`present?` is redundant here.

yuki24 replied:
:+1: I'll remove it.

yuki24 replied:
updated.

vipulnsward commented:
@yuki24 Thanks a ton for this, I was looking to implement something similar due to this - https://twitter.com/vipulnsward/status/603866969201373184

yuki24 commented:
I believe everything should be good and we are ready to merge.

schneems commented:
I spoke to @jeremy, I think it just fell off his radar. Can you rebase, this is good to merge.

jeremy commented:
:+1: Looks like a changelog conflict only, no need to rebase.

yuki24 commented:
@schneems @jeremy Thanks! :heart: :heart_decoration: :yellow_heart: :blue_heart:  :heart: :heart_decoration: :yellow_heart: :blue_heart:

kaspth commented:
@yuki24 nice! Have you sent a pull request to enable `public_file_server` as mentioned in the outdated diffs yet? 😁

yuki24 commented:
@kaspth Not yet. I'm not sure if I can work on it anytime soon, so feel free to send a PR if you could.

schneems commented:
What was the idea?

kaspth commented:
@yuki24 @schneems added a pull request at #22173 :grin:

kaspth commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,22 @@
+*   Deprecate `config.static_cache_control` in favor of
```
Feels like this should be in the Railties CHANGELOG file. I can change it at #22173.
